crypto/src/random/random_coin.rs

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+use crate::{errors::RandomCoinError, Hasher};
+use math::{FieldElement, StarkField};
+use utils::collections::Vec;
+
+// RANDOM COIN TRAIT
+// ================================================================================================
+
+/// Pseudo-random element generator for finite fields.
+///
+/// A random coin can be used to draws elements uniformly at random from the specified base field
+// (which is specified via the `B` type parameter) or from any extension of the base field.
+///
+/// Internally we use a cryptographic hash function (which is specified via the `H` type parameter),
+/// to draw elements from the field. The coin works roughly as follows:
+/// - The internal state of the coin consists of a `seed` and a `counter`. At instantiation
+///   time, the `seed` is set to a hash of the provided bytes, and the `counter` is set to 0.
+/// - To draw the next element, we increment the `counter` and compute hash(`seed` || `counter`).
+///   If the resulting value is a valid field element, we return the result; otherwise we try
+///   again until a valid element is found or the number of allowed tries is exceeded.
+/// - We can also re-seed the coin with a new value. During the reseeding procedure, the
+///   seed is set to hash(`old_seed` || `new_seed`), and the counter is reset to 0.
```

Comment:
I would probably remove it as concrete implementations may chose to do this differently.